flintrock.py

```diff
@@ -526,42 +526,98 @@ def launch_ec2(
+                security_group_ids=[sg.id for sg in security_groups],
+                subnet_id=subnet_id,
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized)
+
+            spot_instance_request_ids = [sir.id for sir in instance_reqs]
+
+            print("Waiting for spot instances to be granted...")
+            try:
+                while True:
+                    time.sleep(30)
+
+                    # Note that there can be outstanding spot instance requests
+                    # that are not part of our current requests (say, from
+                    # another process), and so we have to filter for just
+                    # our requests
+                    all_account_reqs = connection.get_all_spot_instance_requests()
+
+                    active_instance_ids = []
+
+                    for r in all_account_reqs:
+                        req_id = r.id
+                        if req_id in spot_instance_request_ids:
+                            if r.state == 'active':
+                                active_instance_ids.append(r.instance_id)
+
+                    if len(active_instance_ids) == (num_instances):
+                        print("All {num_instances} granted".format(num_instances=num_instances))
+                        reservation_resultset = connection.get_all_reservations(active_instance_ids)
+
+                        all_instances = []
+                        for reservation in reservation_resultset:
+                            all_instances += reservation.instances
+                        break
+                    else:
+                        print("{grant} of {req} instances granted, waiting longer".format(
+                            grant=len(active_instance_ids), req=num_instances))
+            except:
+                print("Canceling spot instance requests")
+                connection.cancel_spot_instance_requests(instance_reqs)
+                # FIXME add warning if any nodes actually launched instances
+
+                sys.exit(1)
+        else:
+            print("Launching {c} instances...".format(c=num_slaves + 1))
+
+            reservation = connection.run_instances(
+                image_id=ami,
+                min_count=(num_slaves + 1),
+                max_count=(num_slaves + 1),
+                key_name=key_name,
+                instance_type=instance_type,
+                placement=availability_zone,
+                security_group_ids=[sg.id for sg in security_groups],
+                subnet_id=subnet_id,
+                placement_group=placement_group,
+                tenancy=tenancy,
+                ebs_optimized=ebs_optimized,
+                instance_initiated_shutdown_behavior=instance_initiated_shutdown_behavior)
+
+            time.sleep(10)  # AWS metadata eventual consistency tax.
+
+            # TODO: Move this to a reusable function and add a limit on
+            #       wait time.
+            while True:
```

Comment:
I mean the wait for the instances to get to a `running` state. We do that so the instances are guaranteed to have been assigned IP addresses, which the rest of the launch code will expect.

Right now we don't explicitly wait for the instances be `running`, so I'm wondering if we should move this block of code outside the `else` clause. Does that make sense?